### ISSUE TITLE
Python implementation?

### ISSUE THREAD

**Issue by jaredballou** [GameServerManagers/LinuxGSM#886]
Hey guys,

I've been poking around at learning Python and using an LGSM build as a way to do so. I've not gotten a whole lot done, as I was mainly working on getting a solid configuration model for game data created. I also need a lot of help figuring out the best ways to package and deploy modules here, as I have never really built a Python framework before. I figured I'd post up my stuff here, and see if I can get some help with the design portion.

My goal with this was to make a simpler and more robust LGSM core that has a very well understood interface. For things like automation, scripting, and the web frontends, I wanted to be able to provide all functionality via command line arguments or configuration files, so that users can deploy it the way they see fit. I am planning to leverage a menu module to provide better user interactive input methods, and to be able to automate every menu or have settings to suppress or answer prompts without interaction.

The repo is at https://github.com/jaredballou/lgsm-python - basically, there are two Python modules in lgsm/modules for lgsmcore (the game script, including script config management and gamedata parser) and steam (installs steamcmd and will install/update games).

The game data is at https://github.com/jaredballou/lgsm-gamedata - they are in a tree where platform, engine, and game are all inherited and interpolated. I am only working on insserver right now since that's the game I care about, so a lot of the other YAML files are either broken or not even YAML (just renamed shell-based files from my LGSM fork). You can get the general idea of what I'm trying to do there to support games without code when possible. I want to be able to include modules from both my LGSM-Python repo and pip itself, any suggestions on how to do that best would be appreciated.

Right now, it's very barebones and basically filed with debug messages and busted functionality. This started out as just a POC to see about how a module might be laid out, and I ended up fleshing it out enough to start to show what I'm trying to do.

There are a few other projects that I a looking at, either from a script functionality standpoint or just having good features that I can leverage.
https://github.com/rzeka/QLDS-Manager - Has a modular approach and interesting way of adding commands and features at runtime.
https://github.com/Holiverh/python-valve - RCON and srcds query module
https://github.com/f0rkz/gameserver-daemon - this is actually a "bad example" I am looking at. It has a lot of hard-coded and duplicated items, isn't very flexible, and isn't very extensible. But it actually works, unlike my tool, so I guess I shouldn't be saying anything :D
https://github.com/Dasister/Source-Query-Class-Python - Simpler Source Query module

That's what I got for now, looking for feedback and help if anyone is interested. I'd be glad to give commit rights to any of you guys, my guess is this is still too newbie-ish to actually be the foundation of the finished product, but it's a good sandbox to prove out and solidify the design. Thanks!

-Jared

**Comment by jaredballou** [comment]
More work done on the develop branch. I merged it all into one big Python module for now, and it load gamedata, displays the installer menu, and gets the appid in response. It also handles script and game config files intelligently, and isn't as recklessly clobber-ific as it was. Tomorrow I'm going to take a crack at getting the actual install/start/stop stuff working in SteamCMD, if anyone wants to take a peek at what I did and let me know what you think I'd appreciate it.
